harmony/harmonydb/sql/20250312-batching-functions.sql

```diff
@@ -0,0 +1,375 @@
+ALTER TABLE sectors_sdr_pipeline
+    ADD COLUMN start_epoch BIGINT DEFAULT NULL;
+
+/*
+  This revised function removes any named temporary tables and instead uses a
+  single CTE chain plus a loop in PL/pgSQL. This way, multiple pollers can run
+  concurrently without hitting errors about existing temp tables.
+
+  Steps:
+    1) Lock all rows needing a commit message:
+         - after_porep = TRUE
+         - porep_proof IS NOT NULL
+         - task_id_commit_msg IS NULL
+         - after_commit_msg = FALSE
+         - start_epoch IS NOT NULL
+       We do so with "FOR UPDATE SKIP LOCKED" to avoid collisions among pollers.
+
+    2) In the same query, number these rows per group (sp_id, reg_seal_proof),
+       chunk them by p_max_batch size, and group them to get each batch's:
+         - sp_id, reg_seal_proof
+         - Array of sector_number
+         - MIN(start_epoch)      as batch_start_epoch
+         - MIN(commit_ready_at)  as earliest_ready_at
+
+    3) Iterate over these batches in ascending order, checking for the first batch
+       that meets ANY of:
+         (a) (batch_start_epoch - p_slack_epoch - p_current_height) <= 0
+         (b) (earliest_ready_at + p_timeout_secs) > now()
+         (c) p_basefee_ok = TRUE
+       If found, update those rows in sectors_sdr_pipeline => task_id_commit_msg = p_new_task_id
+       and return the number of rows updated.
+
+    4) If no batch meets a condition, return 0.
+
+  Usage example:
+    SELECT poll_start_batch_commit_msg(
+      slack_epoch_value,
+      current_height_value,
+      max_batch_size,
+      new_task_id_value,
+      basefee_ok_boolean,
+      timeout_secs
+    );
+
+  Important:
+    - Call this function in a transaction.
+    - Because we do "FOR UPDATE SKIP LOCKED," multiple pollers do not process
+      the same rows simultaneously.
```

Comment:
Pretty sure `FOR UPDATE SKIP LOCKED` doesn't really do anything in Yugabyte due to the default serializable transactions

Reply:
Good catch! I kind of forgot that YB has read isolation = false by default. Will remove these. The function should cover any conflict during update any way.